agent/consul/autopilot/autopilot.go

```diff
@@ -0,0 +1,457 @@
+	if !ok {
+		return nil
+	}
+	// Bail early if autopilot config hasn't been initialized yet
+	if autopilotConf == nil {
+		return nil
+	}
+
+	// Get the the serf members which are Consul servers
+	serverMap := make(map[string]*metadata.Server)
+	for _, member := range a.delegate.Serf().Members() {
+		if member.Status == serf.StatusLeft {
+			continue
+		}
+
+		valid, parts := metadata.IsConsulServer(member)
+		if valid {
+			serverMap[parts.ID] = parts
+		}
+	}
+
+	raftNode := a.delegate.Raft()
+	future := raftNode.GetConfiguration()
+	if err := future.Error(); err != nil {
+		return fmt.Errorf("error getting Raft configuration %s", err)
+	}
+	servers := future.Configuration().Servers
+
+	// Fetch the health for each of the servers in parallel so we get as
+	// consistent of a sample as possible. We capture the leader's index
+	// here as well so it roughly lines up with the same point in time.
+	targetLastIndex := raftNode.LastIndex()
+	var fetchList []*metadata.Server
+	for _, server := range servers {
+		if parts, ok := serverMap[string(server.ID)]; ok {
+			fetchList = append(fetchList, parts)
+		}
+	}
+	d := time.Now().Add(a.healthInterval / 2)
+	ctx, cancel := context.WithDeadline(context.Background(), d)
+	defer cancel()
+	fetchedStats := a.delegate.FetchStats(ctx, fetchList)
+
+	// Build a current list of server healths
+	leader := raftNode.Leader()
+	var clusterHealth OperatorHealthReply
+	voterCount := 0
+	healthyCount := 0
+	healthyVoterCount := 0
+	for _, server := range servers {
+		health := ServerHealth{
+			ID:          string(server.ID),
+			Address:     string(server.Address),
+			Leader:      server.Address == leader,
+			LastContact: -1,
+			Voter:       server.Suffrage == raft.Voter,
+		}
+
+		parts, ok := serverMap[string(server.ID)]
+		if ok {
+			health.Name = parts.Name
+			health.SerfStatus = parts.Status
+			health.Version = parts.Build.String()
+			if stats, ok := fetchedStats[string(server.ID)]; ok {
+				if err := a.updateServerHealth(&health, parts, stats, autopilotConf, targetLastIndex); err != nil {
+					a.logger.Printf("[WARN] autopilot: Error updating server %s health: %s", fmtServer(server), err)
+				}
+			}
+		} else {
+			health.SerfStatus = serf.StatusNone
+		}
+
+		if health.Voter {
+			voterCount++
+		}
+		if health.Healthy {
+			healthyCount++
+			if health.Voter {
+				healthyVoterCount++
+			}
+		}
+
+		clusterHealth.Servers = append(clusterHealth.Servers, health)
+	}
+	clusterHealth.Healthy = healthyCount == len(servers)
+
+	// If we have extra healthy voters, update FailureTolerance
+	requiredQuorum := voterCount/2 + 1
+	if healthyVoterCount > requiredQuorum {
+		clusterHealth.FailureTolerance = healthyVoterCount - requiredQuorum
+	}
+
+	// Heartbeat a metric for monitoring if we're the leader
+	if raftNode.State() == raft.Leader {
+		metrics.SetGauge([]string{"consul", "autopilot", "failure_tolerance"}, float32(clusterHealth.FailureTolerance))
```

Comment:
These still have `consul` in metrics names, will need to parameterize that (I don't think we can take these names away until a major release).

Reply:
You could pass cluster health into a delegate method and put the metrics on the Consul side.